controller/aks-cluster-config-handler.go

```diff
@@ -498,386 +466,132 @@ func BuildUpstreamClusterState(ctx context.Context, secretsCache wranglerv1.Secr
-	}
-
-	logrus.Infof("Resource group %s for cluster [%s] removed successfully", spec.ResourceGroup, spec.ClusterName)
-	return nil
-}
-
-func newClustersClient(authorizer autorest.Authorizer, spec aksv1.AKSClusterConfigSpec) (*containersv211.ManagedClustersClient, error) {
-	if authorizer == nil {
-		return nil, fmt.Errorf("authorizer must not be nil")
-	}
-
-	baseURL := spec.BaseURL
-	if baseURL != nil {
-		baseURL = &azure.PublicCloud.ResourceManagerEndpoint
-	}
-
-	client := containersv211.NewManagedClustersClientWithBaseURI(*baseURL, spec.SubscriptionID)
-	client.Authorizer = authorizer
-
-	return &client, nil
-}
-
-func newServiceClient(authorizer autorest.Authorizer, spec aksv1.AKSClusterConfigSpec) (*containersv209.ContainerServicesClient, error) {
-	if authorizer == nil {
-		return nil, fmt.Errorf("authorizer must not be nil")
-	}
-
-	baseURL := spec.BaseURL
-	if baseURL != nil {
-		baseURL = &azure.PublicCloud.ResourceManagerEndpoint
-	}
-
-	client := containersv209.NewContainerServicesClientWithBaseURI(*baseURL, spec.SubscriptionID)
-	client.Authorizer = authorizer
-
-	return &client, nil
-}
-
-// createOrUpdateCluster creates a new managed Kubernetes cluster
-func createOrUpdateCluster(ctx context.Context, secretsCache wranglerv1.SecretCache, clusterClient *containersv211.ManagedClustersClient, spec aksv1.AKSClusterConfigSpec) (c containersv211.ManagedCluster, err error) {
-
-	dnsPrefix := spec.DNSPrefix
-	if dnsPrefix == nil {
-		dnsPrefix = to.StringPtr(spec.ClusterName)
-	}
-
-	tags := make(map[string]*string)
-	for key, val := range spec.Tags {
-		if val != "" {
-			tags[key] = to.StringPtr(val)
-		}
-	}
-	displayName := spec.ClusterName
-	if displayName == "" {
-		displayName = spec.ClusterName
+		return err
 	}
-	tags["displayName"] = to.StringPtr(displayName)
 
-	kubernetesVersion := spec.KubernetesVersion
-
-	var vmNetSubnetID *string
-	networkProfile := &containersv211.NetworkProfile{}
-	if hasCustomVirtualNetwork(spec) {
-		virtualNetworkResourceGroup := spec.ResourceGroup
-
-		//if virtual network resource group is set, use it, otherwise assume it is the same as the cluster
-		if spec.VirtualNetworkResourceGroup != nil {
-			virtualNetworkResourceGroup = *spec.VirtualNetworkResourceGroup
-		}
-
-		vmNetSubnetID = to.StringPtr(fmt.Sprintf(
-			"/subscriptions/%v/resourceGroups/%v/providers/Microsoft.Network/virtualNetworks/%v/subnets/%v",
-			spec.SubscriptionID,
-			virtualNetworkResourceGroup,
-			spec.VirtualNetwork,
-			spec.Subnet,
-		))
-
-		networkProfile.DNSServiceIP = spec.NetworkDNSServiceIP
-		networkProfile.DockerBridgeCidr = spec.NetworkDockerBridgeCIDR
-		networkProfile.ServiceCidr = spec.NetworkServiceCIDR
-
-		if spec.NetworkPlugin != nil {
-			networkProfile.NetworkPlugin = containersv211.Kubenet
+	updateNodePool := false
+	// check for updated NodePools
+	upstreamNodePools := utils.BuildNodePoolMap(upstreamSpec.NodePools)
+	for _, np := range spec.NodePools {
```

Comment:
This https://github.com/rancher/aks-operator/pull/3#discussion_r616239775 is not addressed yet, is it planned for this PR or will it be addressed in another one? fwiw my comment was based on experimenting with it, not just a guess.

Reply:
Yes, let's do this in another PR. For tracking I have added it to https://github.com/rancher/rancher/issues/31648#issuecomment-829946108